libs/core/execution_base/include/hpx/execution_base/any_sender.hpp

```diff
@@ -0,0 +1,730 @@
+//  Copyright (c) 2021 ETH Zurich
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/assert.hpp>
+#include <hpx/errors/error.hpp>
+#include <hpx/errors/throw_exception.hpp>
+#include <hpx/execution_base/sender.hpp>
+
+#include <cstddef>
+#include <cstring>
+#include <exception>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+namespace hpx::detail {
+    template <typename T>
+    struct empty_vtable
+    {
+        static_assert(
+            sizeof(T) == 0, "No empty vtable type defined for given type T");
+    };
+
+    template <typename T>
+    struct get_empty_vtable
+    {
+        using empty_vtable_type = typename empty_vtable<T>::type;
+        static_assert(std::is_base_of_v<T, empty_vtable_type>,
+            "Given empty vtable type should be a base of T");
+
+        static T* call()
+        {
+            static empty_vtable_type empty;
+            return &empty;
```

Comment:
This is dangerous on platforms that instantiate global and function-local static variables for every module they are visible to (e.g. Windows). In this case, the address returned from `call()` may not be unique.

Reply:
I feared that. Do we need to do the same thing we do for the extra archive data stuff? Explicitly instantiate this for types we need?

Reply:
Does the `empty_table` have to be a template?

Reply:
With the current setup yes, unless I'm misunderstanding the question. What alternatives do you see?

Reply:
I thought that all `any_sender` instances could share the same empty table as in the end all you need is its address. Making `empty_table` a non-template would allow to separate the `call` function into a source file.

Reply:
I see now. There's an empty vtable for `any_sender`, `unique_any_sender`, `any_receiver`, and `any_operation_state`, and I access it generically through this template in the base classes. That's why it's templated. I think I'll just have to explicitly instantiate them in a cpp file.

Reply:
But, but... do we really have to use different instance of the empty table for those?

Reply:
Well... I don't know? I suppose this is where my inexperience with type-erasure shows. With the current setup they do need different instances. `empty_any_operation_state` does not have the same member functions as `empty_any_receiver`. How is this typically dealt with?

In either case, the function static empty vtable(s) have(/has) to go in a cpp file.

Reply:
Hmmm, I might miss something, sorry. If the set of types the empty table has to be instantiated for is confined, then yes, simply explicitly instantiate those in a cpp file and export the corresponding `call()` function.

Reply:
Alternatively you could have separate empty table types (non-templates) for each of the variants and explicitly instantiate the `get_empty_vtable<T>::call()` function

Reply:
> Hmmm, I might miss something, sorry.

So may I.

The set of types is confined at least in this context. If `sbo_storage` is to be useful elsewhere it's of course not confined anymore, though I wouldn't put that in our public API in any case. The empty vtable types themselves are not templated. It's just that I access the (four) different empty vtable types (their instances) through `get_empty_vtable` in `sbo_storage`. It could actually also be a function template, but that doesn't matter too much.

Reply:
Hmm, I take back the part about the set of types being confined. The base types for senders and receivers (not operation state) are templated on the types sent by the wrapped sender. Just to understand the restrictions we're dealing with, this is also problematic, right? https://github.com/STEllAR-GROUP/hpx/blob/65280c318d97c53bee082b8ce87ce78922660070/libs/parallelism/execution/include/hpx/execution/executors/polymorphic_executor.hpp#L592-L599 (it's similarly templated on a function signature which is provided by the user). If that's fine then this would be fine as well? If it's not, then whatever fix I end up with here should be applied there as well.

Reply:
> Hmm, I take back the part about the set of types being confined. The base types for senders and receivers (not operation state) are templated on the types sent by the wrapped sender. Just to understand the restrictions we're dealing with, this is also problematic, right?
> 
> https://github.com/STEllAR-GROUP/hpx/blob/65280c318d97c53bee082b8ce87ce78922660070/libs/parallelism/execution/include/hpx/execution/executors/polymorphic_executor.hpp#L592-L599
> 
> (it's similarly templated on a function signature which is provided by the user). If that's fine then this would be fine as well? If it's not, then whatever fix I end up with here should be applied there as well.

Yes, this is equally problematic. Good catch.

Reply:
Currently the only reasonable solution I can think of is representing the empty state by `nullptr` rather than a pointer to a base class, and hope that `if (HPX_LIKELY(ptr)) ptr->member(...); else empty{}.member(...);` is not too expensive. Branch prediction *should* make it cheap in any case, but it's not quite the same as avoiding the branch in the first place.

Reply:
We discussed this in the meeting and there are essentially two approaches around this:
- represent the empty state with nullptr and pay for a branch on dispatch
- store a boolean to represent the empty state, and don't pay for branches on dispatch
I've implemented the second here. I think the polymorphic executor is actually fine after looking at it again. It also uses extra storage to represent the empty state, and keeps dispatch branch-free.

Reply:
A third option would be to add another function to the vtable that returns a bool. This safes on memory for the `any_sender` but adds a slight overhead to the check for emptyness (indirect function call instead of simple memory read).

Reply:
I've updated it to use the third option.